The `sparse_attention` operator fails the benchmark test.

### Describe the bug

The `sparse_attention` operator fails benchmark testing.
The output is as follows:

```

                             "(`_builder` argument must be provided outside of JIT functions.)")
>       return fn(*args, **kwargs)

/usr/local/lib/python3.11/dist-packages/triton/language/core.py:34: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/usr/local/lib/python3.11/dist-packages/triton/language/core.py:1814: in dot
    return semantic.dot(input, other, acc, input_precision, max_num_imprecise_acc, out_dtype, _builder)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

lhs = <triton.language.core.tensor object at 0x7efaedd45890>
rhs = <triton.language.core.tensor object at 0x7efaedd4cfd0>, acc = None
input_precision = <INPUT_PRECISION.TF32: 0>, max_num_imprecise_acc = None
out_dtype = triton.language.float32
builder = <triton._C.libtriton.ir.builder object at 0x7efaede99850>

    def dot(lhs: tl.tensor, rhs: tl.tensor, acc: tl.tensor, input_precision: Optional[str], max_num_imprecise_acc: int,
            out_dtype: tl.dtype, builder: ir.builder) -> tl.tensor:
        assert lhs.type.is_block() and rhs.type.is_block()
    
        if lhs.dtype.is_fp8() and rhs.dtype.is_fp8():
            # All combinations of supported fp8 x fp8 are permitted
            pass
        else:
            assert lhs.dtype in (tl.int8, tl.uint8, tl.float16, tl.bfloat16,
                                 tl.float32), f"Unsupported lhs dtype {lhs.dtype}"
            assert rhs.dtype in (tl.int8, tl.uint8, tl.float16, tl.bfloat16,
                                 tl.float32), f"Unsupported rhs dtype {rhs.dtype}"
            assert lhs.dtype == rhs.dtype, f"Both operands must be same dtype. Got {lhs.dtype} and {rhs.dtype}"
    
        if lhs.dtype.is_fp8e4b15() or rhs.dtype.is_fp8e4b15():
            # We upcast because there's no fp8e4b15 type in MLIR
            lhs = cast(lhs, tl.float16, builder)
            rhs = cast(rhs, tl.float16, builder)
    
        if input_precision is None:
            input_precision = builder.options.default_dot_input_precision
    
        input_precision = _str_to_dot_input_precision(input_precision, builder)
    
        lhs_rank = len(lhs.shape)
        rhs_rank = len(rhs.shape)
        assert lhs_rank == rhs_rank == 2 or lhs_rank == rhs_rank == 3, f"Both inputs must be either 2D or 3D; (lhs: {lhs.shape} vs rhs: {rhs.shape})"
        assert lhs.shape[-1].value == rhs.shape[
            -2].value, f"First input shape ({lhs.shape}) and second input shape {rhs.shape} are not compatible for matmul (second index of first shape ({lhs.shape[-1].value}) must be equal to first index of second shape ({rhs.shape[-2].value})"
        assert builder.codegen_fns.get("min_dot_size") is not None, "target doesn't provide lower shape bounds for dot."
        min_dot_size = builder.codegen_fns["min_dot_size"](lhs.type, rhs.type)
>       assert lhs.shape[-2].value >= min_dot_size[0] and lhs.shape[-1].value >= min_dot_size[2] \
            and rhs.shape[-1].value >= min_dot_size[1], \
                f"Input shapes should have M >= {min_dot_size[0]}, N >= {min_dot_size[1]} and K >= {min_dot_size[2]}"
E       AssertionError: Input shapes should have M >= 16, N >= 16 and K >= 16

/usr/local/lib/python3.11/dist-packages/triton/language/semantic.py:1570: AssertionError

```

--- a/benchmark/test_sparse_attention.py
+++ b/benchmark/test_sparse_attention.py
@@ -69,7 +69,6 @@ class SparseAttentionBenchmark(base.Benchmark):
             yield q, kv, attn_sink, topk_idxs, 1.0 / math.sqrt(dim)
 
 
-@pytest.mark.skip(reason="The test case fails with many exceptions: #2669")
 @pytest.mark.skipif(flag_gems.device == "cpu", reason="Unsupported in CPU mode")
 @pytest.mark.sparse_attn_triton
 def test_sparse_attn_triton():

--- a/src/flag_gems/fused/sparse_attention.py
+++ b/src/flag_gems/fused/sparse_attention.py
@@ -31,6 +31,7 @@ def sparse_attn_triton_kernel(
     stride_idxk,
     scale,
     topk,
+    H_ACTUAL,
     BLOCK: tl.constexpr,
     D: tl.constexpr,
     H: tl.constexpr,
@@ -42,8 +43,9 @@ def sparse_attn_triton_kernel(
     q_base = Q + pid_b * stride_qb + pid_m * stride_qm
     offs_h = tl.arange(0, H)
     offs_d = tl.arange(0, D)
+    h_mask = offs_h < H_ACTUAL
     q_ptrs = q_base + offs_h[:, None] * stride_qh + offs_d[None, :] * stride_qd
-    q_block = tl.load(q_ptrs)  # (H, D) bf16
+    q_block = tl.load(q_ptrs, mask=h_mask[:, None], other=0.0)  # (H, D) bf16
 
     # ---- base pointers ----
     kv_base = KV + pid_b * stride_kvb
@@ -95,7 +97,7 @@ def sparse_attn_triton_kernel(
         sum_exp = sum_exp * correction + scores_sum
 
     # ---- incorporate attn_sink ----
-    sink_vals = tl.load(attn_sink + offs_h)  # (H,)
+    sink_vals = tl.load(attn_sink + offs_h, mask=h_mask, other=0.0)  # (H,)
     sum_exp = sum_exp + tl.exp(sink_vals - scores_max)
 
     # ---- normalize ----
@@ -104,7 +106,7 @@ def sparse_attn_triton_kernel(
     # ---- store output: (H, D) ----
     o_base = O + pid_b * stride_ob + pid_m * stride_om
     o_ptrs = o_base + offs_h[:, None] * stride_oh + offs_d[None, :] * stride_od
-    tl.store(o_ptrs, acc_o.to(tl.bfloat16))
+    tl.store(o_ptrs, acc_o.to(tl.bfloat16), mask=h_mask[:, None])
 
 
 # ---------------------------------------------------------------------------
@@ -121,6 +123,7 @@ def sparse_attn_triton(
     topk = topk_idxs.shape[-1]
     o = torch.empty_like(q)
     BLOCK = 64
+    h_padded = max(16, triton.next_power_of_2(h))
 
     grid = (m, b)  # each program handles ALL h heads
     sparse_attn_triton_kernel[grid](
@@ -145,9 +148,10 @@ def sparse_attn_triton(
         topk_idxs.stride(2),
         softmax_scale,
         topk,
+        h,
         BLOCK=BLOCK,
         D=d,
-        H=h,
+        H=h_padded,
         num_warps=8,  # 256 threads, matching tilelang
     )
     return o

--- a/src/flag_gems/fused/sparse_attention.py
+++ b/src/flag_gems/fused/sparse_attention.py
@@ -77,9 +77,8 @@ def sparse_attn_triton_kernel(
         # -- scores: Q @ KV^T -> (H, BLOCK) via GEMM --
         acc_s = tl.dot(q_block, tl.trans(kv_block))  # (H, D) @ (D, BLOCK) = (H, BLOCK)
         acc_s = acc_s * scale
-        # mask invalid positions to -inf
-        mask_bias = tl.where(valid_mask, 0.0, float("-inf"))  # (BLOCK,)
-        acc_s = acc_s + mask_bias[None, :]  # broadcast: (H, BLOCK)
+        # mask invalid positions to -inf (apply on 2D tensor to avoid layout mismatch)
+        acc_s = tl.where(valid_mask[None, :], acc_s, float("-inf"))
 
         # -- online softmax update --
         scores_max_prev = scores_max

--- a/tests/test_flash_attention.py
+++ b/tests/test_flash_attention.py
@@ -160,9 +160,6 @@ def sparse_attention_ref(q, kv, attn_sink, topk_idxs, scale):
     return out.to(q.dtype)
 
 
-@pytest.mark.skip(
-    reason="Issue #2809: The operator fails this test on Nvidia at least."
-)
 @pytest.mark.skipif(cfg.TO_CPU, reason="Unsupported in CPU mode")
 @pytest.mark.sparse_attn_triton
 @pytest.mark.parametrize(

